### PR TITLE
adding burst class

### PR DESCRIPTION
**Pull request by RobinGuillard**
First step in episode analysis. Defining burst instances with a beggining, an end, the possibility to mark them as tonic and basic functions to manipulate them.

### CODE REVIEW

**Comment by lkorczowski** [comment]
@RobinGuillard please remove `Notebooks/Bruxism_detection_vbrouillon.ipynb` I don't even understand how it is still there :/

**Comment by RobinGuillard** [comment]
Please review the present new branch.
Here are a few explanations of my code:
Starting with a classification result (list or booleans), in order to assess where are the bruxism episodes and what type they are (tonic, phasic or mixed), we need to have a 2 step approach. First we need to "merge" the contiguous bursts (series of True value in the classification result) into single bursts.
This is what the classif_to_burst function does (in the scoring.py code), thanks to the "burst" class.
Then, having in a list all the merged burst, we need to assess where are the real "episodes" of bruxism according to the medical consensus definition. Isolated short bursts should be deleted and only valid episodes that are either tonic, phasic or mixed should remain. For further analysis, it can be interesting though to keep in memory which bursts are contained within each episode.
This is what the burst_to_episode function of scoring does, with the help of the episode class.

At last, we want to be able to compare quickly the scoring of an expert with the automatic scoring, this is why I have made the function create_list_events in scoring.py

Burst and episode classes are here to simplify the scoring.py functions and deal with special cases (overlapping etc...) which will occur rarely.

Surely the code I developed is too much for the issue to handle, I think I can simplify greatly the add_a_burst function tomorrow. Please Take the time to consider that creating the episode labels is not as easy as eit seems before thinking I've done it wrong and starting again from scratch

**Comment by RobinGuillard** [comment]
I corrected one error in test_episode, it should work better, and maybe pass the tests!

**Comment by lkorczowski** [comment]
`Notebooks/Bruxism_detection_vbrouillon.ipynb → draft/Bruxism_detection_vbrouillon.ipynb`

The gitignore folder is `draft*s*/`

Carefull ^^

**Comment by RobinGuillard** [comment]
I made again a pytest trial this morning, and just because I opened up a new terminal, your function "check_if_fitted" began again to throw an error. I'll make an issue out of it. By removing it, it passes all the tests!

**Comment by RobinGuillard** [comment]
Ok I really don't know why I don't pass the tests, everything is running fine on my side. and when I click on accessing the codecov report, I get access denied. I'll need you to understand what is happening here ^^

**Comment by lkorczowski** [comment]
If you fail codecov coverage, find the non-tested lines by doing in terminal:
```
coverage run -m pytest
coverage report -m
```

the non-covered lines will be mentioned. You could also use PyCharm Professional which has an in-build coverage feature (i.e. non-covered lines are highlighted in your editor)

File: tinnsleep/burst.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            if (self.end - self.beg > 2):
+                self.is_tonic=True
+            else:
+                self.is_tonic=False
+            
+            #print(self.beg)
+            #print(self.end)
+            return  True
+        else:
+            return False
```
[the remark this answers is not in the document]
Comment (RobinGuillard) [reply tinnsleep/burst.py:103]: done :)

File: tinnsleep/classification.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ def _predict_distances(self, X):
     def predict(self, X):
 
         # Check is fit had been called
-        check_is_fitted(self)
+        #check_is_fitted(self)
```
[the remark this answers is not in the document]
Comment (RobinGuillard) [reply tinnsleep/classification.py:72]: I deleted the line

File: tinnsleep/episode.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+
+    See Also
+    --------
+
+    """
+
+    def __init__(self, burst):
+        self.beg = burst.beg
+        self.end = burst.end
```
[the remark this answers is not in the document]
Comment (RobinGuillard) [reply tinnsleep/episode.py:28]: Not sure that episode is directly inheriting from burst as they have different initialisation functions, notably regarding the episode.burst_list attribute. You cannot create an episode the same way you create a burst

File: tinnsleep/episode.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            #Check if it is now not overlapping with the one before
+            if self.nb_burst>1:
+                if self.burst_list[-2].merge_if_overlap(self.burst_list[-1]):
+                                self.burst_list.pop(-1)
+                                self.nb_burst-=1
+        #Case if it is not overlapping
+        elif self.end<=burst.beg:
+            self.end = burst.end
+            self.burst_list.append(burst)
+            self.nb_burst+=1
```
[the remark this answers is not in the document]
Comment (RobinGuillard) [reply tinnsleep/episode.py:56]: Totally agreeing with you, Suppressing the nb_burst object...

**Comment by RobinGuillard** [comment]
I have added some corrections:
- Now burst can be created with end < beg
- Now burst_to_episode works in all cases (thanks to new rearrange_chronological function)
- There was one case of add_a_burst that wasn't covered (Second episode addition with two borns modification). It is now fixed and tested.

File: tinnsleep/scoring.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
+from tinnsleep.burst import burst
+from tinnsleep.episode import episode
+
+def classif_to_burst(classif, interval=0.25):
```
[the remark this answers is not in the document]
Comment (RobinGuillard) [reply tinnsleep/scoring.py:5]: Oh yes sorry!

[the remark this answers is not in the document]
Comment (RobinGuillard) [reply tinnsleep/scoring.py:5]: But the thing is: if I put an interval in number of samples, then I have to add the s_freq input as well. We need a time interval of 0.25! I keep the time interval but I will name it time_interval for easy differentiation.

**Comment by RobinGuillard** [comment]
I still don't get why the coverage tests are not working + despite your explanations, I couldn't run
```
coverage run -m pytest
coverage report -m
```

On my computer. I don't know why it is not working anymore... Another fix to make...

File: tests/test_episode.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+from tinnsleep.burst import burst
+from tinnsleep.episode import episode
+
+
+def test_episode():
+    bursty = burst(0.1, 1)
+    epi = episode(bursty)
+    npt.assert_equal(epi.beg, 0.1)
+    npt.assert_equal(epi.end, 1)
+    npt.assert_equal(len(epi.burst_list), 1)
```
Comment (lkorczowski) [inline tests/test_episode.py:12]: replace by:
```
npt.assert_equal([epi.beg, epi.end, len(epi.burst_list), [0.1, 1, 1])
```

apply that comment to all multi `npt.assert_equal`

File: tests/test_episode.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+def test_overlap_left():
+    bursty = burst(1.5, 2.5)
+    bursty2 = burst(1.75, 3.5)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
Comment (lkorczowski) [inline tests/test_episode.py:21]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    bursty2 = burst(4.75, 5.5)
+    bursty3 = burst(3, 3.9)
+    bursty4 = burst(2, 3.4)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty4)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 5.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
Comment (lkorczowski) [inline tests/test_episode.py:34]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def test_no_overlap_left():
+    bursty = burst(1.5, 2.5)
+    bursty2 = burst(2.75, 3.5)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
Comment (lkorczowski) [inline tests/test_episode.py:44]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def test_overlap_right():
+    bursty = burst(1.5, 2.5)
+    bursty2 = burst(1.75, 3.5)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
Comment (lkorczowski) [inline tests/test_episode.py:54]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+def test_overlap_right_2_bursts():
+    bursty = burst(1.5, 2.5)
+    bursty3 = burst(2.6, 2.9)
+    bursty2 = burst(2.75, 3.5)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty2)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
Comment (lkorczowski) [inline tests/test_episode.py:65]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+def test_overlapx2_right():
+    bursty = burst(1.5, 2.5)
+    bursty3 = burst(2, 2.9)
+    bursty2 = burst(2.75, 3.5)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    epi.add_a_burst(bursty3)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
Comment (lkorczowski) [inline tests/test_episode.py:77]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def test_no_overlap_right():
+    bursty = burst(1.5, 2.5)
+    bursty2 = burst(2.75, 3.5)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
Comment (lkorczowski) [inline tests/test_episode.py:87]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+def test_megaburst():
+    bursty = burst(1.5, 2.5)
+    bursty2 = burst(1.75, 3.5)
+    bursty3 = burst(1.4, 3.6)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    epi.add_a_burst(bursty3)
+    npt.assert_equal(epi.beg, 1.4)
+    npt.assert_equal(epi.end, 3.6)
+    npt.assert_equal(len(epi.burst_list), 1)
```
Comment (lkorczowski) [inline tests/test_episode.py:99]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def test_burst_already_there():
+    bursty = burst(1.5, 2.5)
+    bursty2 = burst(1.5, 2.5)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 2.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
Comment (lkorczowski) [inline tests/test_episode.py:109]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    bursty2 = burst(4.75, 5.5)
+    bursty3 = burst(3, 3.9)
+    bursty4 = burst(2.7, 3.9)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty4)
+    npt.assert_equal(epi.burst_list[1].beg, 2.7)
+    npt.assert_equal(epi.burst_list[1].end, 3.9)
+    npt.assert_equal(len(epi.burst_list), 3)
```
Comment (lkorczowski) [inline tests/test_episode.py:122]: multi `npt.assert_equal`

File: tests/test_episode.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    bursty4 = burst(2.7, 2.9)
+    bursty5 = burst(2.8, 3.7)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty4)
+    epi.add_a_burst(bursty5)
+    npt.assert_equal(epi.burst_list[1].beg, 2.7)
+    npt.assert_equal(epi.burst_list[1].end, 3.9)
+    npt.assert_equal(len(epi.burst_list), 3)
```
Comment (lkorczowski) [inline tests/test_episode.py:137]: multi `npt.assert_equal`

**Comment by RobinGuillard** [comment]
Ok It's done!! Moreover, I successfully set up coverage on pycharm, it's working fine :)

**Pull request merged**